### PR TITLE
Newsletter

### PR DESCRIPTION
**Pull request by thesabinelim**
Implemented ?newsletter command to sign user up for the society newsletter, pulling their details including email from the database.
Added newsletter URL setting in config.

### CODE REVIEW

File: iam/hooks.py [hunk 1]
```diff
@@ -257,9 +257,12 @@ def was_verified_user(cog, obj, *func_args, **func_kwargs):
...
-        return
-    member_data = cog.db.get_member_data(obj.id)
-    if member_data is None or not member_data[MemberKey.ID_VER]:
+        return True, None
+    try:
+        member_data = cog.db.get_member_data(obj.id)
+    except MemberNotFound:
+        return False, "You must be verified to do that."
+    if not member_data[MemberKey.ID_VER]:
         return False, "You must be verified to do that."
```
Comment (TerryAg) [inline iam/hooks.py:266]: I never like functions returning more than one thing like this. Generally if you have to do this, you should be creating a custom class that has both of these values as attributes.

File: iam/hooks.py [hunk 2]
```diff
@@ -285,6 +288,36 @@ def is_unverified_user(cog, obj, *func_args, **func_kwargs):
         return False, "You are already verified."
     return True, None
 
+def is_strictly_verified_user(cog, obj, *func_args, **func_kwargs):
```
Comment (TerryAg) [inline iam/hooks.py:291]: these should just be `*args` and `**kwargs`

File: iam/hooks.py (hunk 1 above)
Comment (TerryAg) [reply iam/hooks.py:266]: In other words, Go != Python

File: iam/hooks.py [hunk 3]
```diff
@@ -285,6 +288,36 @@ def is_unverified_user(cog, obj, *func_args, **func_kwargs):
...
+
+    Returns:
+        1. Boolean result of check.
+        2. Error message to supply, if check failed.
+    """
+    if not isinstance(obj, Member):
+        obj = obj.author
+    err_msg = "You must be verified to do that."
+    member = get_member(cog.bot, obj)
+    if member is not None and VER_ROLE in get_role_ids(member):
```
Comment (TerryAg) [inline iam/hooks.py:310]: sidenode: "VER_ROLE" is kinda  a bad variable name. Verified role? Version role?

Comment (TerryAg) [inline iam/hooks.py:310]: What is the point of this if condition if you have the try/except later? Ditch the if statement. Ask for forgiveness not permission. Avoid race conditions.

File: iam/hooks.py (hunk 2 above)
Comment (TerryAg) [inline iam/hooks.py:291]: What is a "strictly verified user"?

File: iam/newsletter.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    """Subscribe a user to the newsletter using their stored email.
+
+    Args:
+        db: Database object.
+        user: User object to subscribe.
+        channel: Channel to send confirmation message to.
+    """
+    member_data = db.get_member_data(user.id)
+    zid = member_data[MemberKey.ZID]
+    get(f"{NEWSLETTER_SUB_URL}"
```
Comment (TerryAg) [inline iam/newsletter.py:55]: This does not look sanitised. Please look into libraries for doing correct URL parsing.

File: iam/verify.py [hunk 5]
```diff
@@ -747,7 +747,9 @@ async def proc_grant_rank(ver_role, admin_channel, member, silent=False):
     await member.add_roles(ver_role)
     LOG.info(f"Granted verified rank to member '{member.id}'")
     if not silent:
-        await member.send("You are now verified. Welcome to the server!")
+        await member.send("You are now verified. Welcome to the server!"
+            "If you are interested in subscribing to our newsletter, try the "
+            f"{PREFIX}newsletter command.")
```
Comment (TerryAg) [inline iam/verify.py:752]: Add some backticks around the `?newsletter`

File: iam/hooks.py (hunk 2 above)
Comment (thesabinelim) [reply iam/hooks.py:291]: A user who has a verified entry in the database, and not just the role in the server.

File: iam/newsletter.py (hunk 4 above)
Comment (thesabinelim) [reply iam/newsletter.py]: Module has been rewritten to use the official Mailchimp library

**Review by TerryAg — APPROVED** [review]

**Pull request merged**